airbyte-lib: Add telemetry

flash1293 opened the pull request:
This PR adds tracking to airbyte-lib.

On start of each sync, an event is sent to segment, tracking the date, the used source, cache and airbyte-lib version.
Once the sync finishes, a second event is sent, tracking whether the sync was successful or not.

This is done using basic http requests.

If the `DO_NOT_TRACK` env variable is set, no tracking events are sent.

This PR also sets the user agent for the connector registry json file request to make it possible to identify airbyte-lib requests.

aaronsteers commented:
Note from @evantahler - "SyncStats" is the object to match.

evantahler commented:
^ SyncStats is the object in the platform that we send to segment today, and then get into BQ and use in our analysis.  If you could match as much of that object as you can, and tag your lib syncs with a new 'platform' name, all the business tables we have should work fine with this new data!

flash1293 commented:
Thanks for the hint @evantahler 

The thing I'm unsure about is the ways this is different from our regular sync stats which seems to make it incompatible in some key ways that make a separate business table the best approach:
* There is no destination connector
* There is no workspace / user identifier
* The source can be run in multiple ways (venv vs. local installation)
* airbyte-lib follows its own versioning which is independent from platform versions

This mixture of additional properties and the absence of properties that are guaranteed to be available for regular sync stats makes me suspect that a separate business table is the way to go.

@alex-gron what's your opinion on this? What would be the best way to handle this new data?

aaronsteers commented on airbyte-lib/airbyte_lib/telemetry.py:
```diff
@@ -0,0 +1,37 @@
+    if number_of_records is not None:
+        payload["properties"]["number_of_records"] = number_of_records
+
+    # Do not handle the response, we don't want to block the execution
+    requests.post("https://api.segment.io/v1/track", auth=(TRACKING_KEY, ""), json=payload)
```
Totally optional, but a common convention in Python if you want to call out that the intentional dropping of a return value - you can use "_ = ..."

This isn't necessary by any means, but will inform the reader "yes, I know there's a return value and no, I don't want to use it." 😄 

```suggestion
    # Do not handle the response, we don't want to block the execution
    _ = requests.post("https://api.segment.io/v1/track", auth=(TRACKING_KEY, ""), json=payload)
```

aaronsteers commented on airbyte-lib/airbyte_lib/source.py:
```diff
@@ -168,7 +170,7 @@ def check(self):
         self.executor.install()
 
-    def _read(self) -> Iterable[AirbyteRecordMessage]:
+    def _read(self, target: str) -> Iterable[AirbyteRecordMessage]:
```
It's not clear to me what 'target' means in this context. Can we add a docstring to explain it?

flash1293 replied:
It's about where the data is loaded to - a cache or streamed directly via iterable.

aaronsteers commented on airbyte-lib/airbyte_lib/source.py:
```diff
@@ -191,9 +193,9 @@ def _read(self) -> Iterable[AirbyteRecordMessage]:
+        yield from self._read_catalog(target, configured_catalog)
 
-    def _read_catalog(self, catalog: ConfiguredAirbyteCatalog) -> Iterable[AirbyteRecordMessage]:
+    def _read_catalog(self, target: str, catalog: ConfiguredAirbyteCatalog) -> Iterable[AirbyteRecordMessage]:
```
Ditto above. I'm not sure what 'target' means here.

aaronsteers commented on airbyte-lib/airbyte_lib/executor.py:
```diff
@@ -28,6 +28,10 @@ def ensure_installation(self):
         pass
 
+    @abstractmethod
+    def get_tracking_information(self) -> Dict[str, str]:
```
This might be semantic, but I slightly prefer a name like `get_telemetry_info` or `get_telemetry_markers` over `get_tracking_info`. The work 'tracking', to me at least, implies ability to personally identify or de-anonymize, which this isn't.

flash1293 replied:
Makes sense to me, I pondered the name a bit. Telemetry has the risk to be confused with telemetry in the sense of OpenTelemetry and the like, but it's probably better than tracking. "analytics" would be another option, but thinking about it calling this kind of thing telemetry is probably an industry standard for software running on the client.

flash1293 commented:
Thanks for the review @aaronsteers - Your comment about the `target` made me think that the current PR was a little underspecified in terms of the data shape of the telemetry.

I adjusted that and introduced data classes and enums for all the bits instead of dicts and strings, feels safer now and a little more "self-documenting".

Using loose types can be really deceiving as the data needs to be shaped correctly to be handled correctly in the downstream processing anyway.

Let me know what you think

alex-gron commented:
> Thanks for the hint @evantahler
> 
> The thing I'm unsure about is the ways this is different from our regular sync stats which seems to make it incompatible in some key ways that make a separate business table the best approach:
> 
> * There is no destination connector
> * There is no workspace / user identifier
> * The source can be run in multiple ways (venv vs. local installation)
> * airbyte-lib follows its own versioning which is independent from platform versions but can overlap
> 
> This mixture of additional properties and the absence of properties that are guaranteed to be available for regular sync stats makes me suspect that a separate business table is the way to go.
> 
> @alex-gron what's your opinion on this? What would be the best way to handle this new data?

@flash1293 A few thoughts here:
* Is it possible for the airbyte-lib to simply use SyncStats since it is in the `airbyte` repo? 
  * If yes, then this is a viable solution and worth considering
  * That said, it is also fine to have these events in a new table, and we can union the data together in dbt. We do this already, and it is not a big lift to add another table. 
* My biggest request is that we add some telemetry for all of the things you mention above (especially venv vs local if that is something you want to track) 
* If in the SyncStats table/object, then we will definitely want a column to indicate that this event is from `airbyte-lib` 
  * There is an existing `airbyte_source` field that we could use for this, but it is currently set to `unknown` a majority of the time, so I'm not 100% sure what it is supposed to contain. 
* Regardless of how we get the raw data (same table or a new one), we can consider adding this data to the Connection Sync business table. 
  * The drivers of that decision will be: what is the granularity of the data and what types of questions do we want to answer about it?

flash1293 commented:
Thanks for the look @alex-gron 

> Is it possible for the airbyte-lib to simply use SyncStats since it is in the airbyte repo?

This is a separate code base and it's not 100% the same data that's provided so we can't reuse on that level.


> My biggest request is that we add some telemetry for all of the things you mention above

The practical things I'm worried about adding this to the existing business table are:
* dbt transforms getting more complex because some columns are not always set (like workspace or destination)
* adding new columns to the existing business table (for local vs. venv and which cache type is used) that are only relevant for airbyte-lib and nothing else

If you feel like these are not really problems, then let's go for that.

aaronsteers commented on airbyte-lib/airbyte_lib/telemetry.py:
```diff
@@ -0,0 +1,63 @@
+    if number_of_records is not None:
+        payload["properties"]["number_of_records"] = number_of_records
+
+    # Do not handle the response, we don't want to block the execution
+    _ = requests.post("https://api.segment.io/v1/track", auth=(TRACKING_KEY, ""), json=payload)
```
Something like this using `contextlib.suppress`, I think:

```suggestion
    # Suppress exceptions if host is unreachable or network is unavailable
    with suppress(OSError):  # Network errors are subclasses of `OSError`
        # Do not handle the response, we don't want to block the execution
        _ = requests.post("https://api.segment.io/v1/track", auth=(TRACKING_KEY, ""), json=payload)
```

flash1293 replied:
Great suggestion!

aaronsteers commented on airbyte-lib/tests/integration_tests/test_integration.py:
```diff
@@ -111,6 +114,77 @@ def test_sync():
+    Test that the telemetry is sent when the sync is successful.
+    This is done by mocking the requests.post method and checking that it is called with the right arguments.
+    """
+    os.environ["DO_NOT_TRACK"] = ""
```
I think we should use a context manager for this override, to prevent bleeding to other tests:

Something like this:

```py
    with patch.dict('os.environ', {'DO_NOT_TRACK': ''}):
        # Your test code here
        # Within this block, 'DO_NOT_TRACK' is set to ''
        # Other environment variables are kept
```

Or if we want `DO_NOT_TRACK` to be _unset_, then maybe something like this:

```py
    with patch.dict('os.environ', values={}, remove=['DO_NOT_TRACK']):
        # Your test code here
        # Within this block, 'DO_NOT_TRACK' is unset
        # Other environment variables are kept
```

I also think we should be clear about whether an empty string still counts as the variable being 'set'. I think this implementation would treat the empty string as 'unset', which makes sense I think. In the implementation files, I was originally going to recommend checking for `"DO_NOT_CHECK" not in os.environ` but that check would not work if we want the empty string to count as not being set.

I'm totally okay with either approach. Just calling out the distinction to make sure we're not missing anything.

aaronsteers replied:
After giving this some more thought and research, I think I'm in the camp of the empty string being 'falsey' (tracking enabled) and any other set value being 'truthy' (meaning opted-out). This is easy to document and communicate, easy to modify, and unlikely to lead to any users _thinking_ they are opted out when they actually aren't.

flash1293 replied:
Good suggestion to use patch.dict here!

Agreed, it depends on the nature of the environment variable, but for something like do_not_track it feels right to err on the side of interpreting something as truthy even if it wasn't meant like that - better than the other way around.

aaronsteers commented on airbyte-lib/airbyte_lib/telemetry.py:
```diff
@@ -0,0 +1,63 @@
+        return
+
+    current_time = datetime.datetime.utcnow().isoformat()
+    payload: Dict[str, Any] = {
+        "anonymousId": "airbyte-lib-user",
+        "event": "sync",
+        "properties": {
```
I think I'd like to request one more attribute here, which is capturing the presence of a "CI" flag in os.environ. 

CI tests are very noisy on our metrics - with the possibility to make up 50-80% of usage stats in some extreme cases, or in specific cross-sections of usage. Luckily, the industry has a convention:

https://chat.openai.com/share/eb500944-8bb6-43a1-b64e-5f901f4587c4

_(Turns out "CI environment variable" is hard to google. I've previously seen a third party site advocating for this standard but I couldn't find it today so the above link is a ChatGPT conversation.)_

Rather than return the value stored in the "CI" environment variable, I think it is sufficient to list it in 'extra_flags' or similar, or just as its own boolean ("True" means present").

Wdyt?

aaronsteers reviewed (approved):
@flash1293 - Thanks for moving this forward! I've added several suggestions, but none need to block this increment. I'm approving with the noted suggestions, and we can follow-up in subsequent iterations if needed.

aaronsteers commented on airbyte-lib/airbyte_lib/telemetry.py:
```diff
@@ -0,0 +1,63 @@
+    LOCAL_INSTALL = "local_install"
+
+
+class CacheType(str, Enum):
+    # TODO: Extend once proper caches have been merged
+    IN_MEMORY = "in_memory"
+    STREAMING = "streaming"
```
Even though it wouldn't contain much, we could also adopt the "TelemetryInfo" convention to be parallel with the source side. In that case, this would be a "CacheTelemetryInfo" dataclass, perhaps just with one member, and would render as a dict instead of a string in the telemetry object.

flash1293 replied:
> Since in-memory is out of scope for now, what do you think of the cache types just being the name of the DB backed? So, something like DUCKDB, POSTGRES, SNOWFLAKE, etc

Yes, that's what the TODO on top is meant for.

> Even though it wouldn't contain much, we could also adopt the "TelemetryInfo" convention to be parallel with the source side. In that case, this would be a "CacheTelemetryInfo" dataclass, perhaps just with one member, and would render as a dict instead of a string in the telemetry object.

Good suggestion, I will adjust it like that - maybe we want to add other dimensions later on.

flash1293 commented on airbyte-lib/airbyte_lib/telemetry.py:
```diff
@@ -0,0 +1,63 @@
+            "source": asdict(source),
+            "state": state,
+            "cache_type": cache_type,
+            # explicitly set to 0.0.0.0 to avoid leaking IP addresses
+            "ip": "0.0.0.0",
```
then segment would automatically add back the actual IP

flash1293 commented:
Thanks for your comments @aaronsteers - I addressed everything and it seems like it's ready to merge but leaving it in place in case you want to take a final look.

alex-gron commented:
@flash1293  Sorry for my delay here. Not sure where you ended up on the same/separate Segment table. 

- To keep things cleaner, I think using a separate Segment table makes sense. Especially since we won't be reusing the existing SyncStats code. 
- Once we have the data in BQ, we can decide if it makes sense to do a separate business table or not. I'm thinking this will probably be similar to the API where we do have a separate table because the granularity is different.

Let me know if that answers your questions and unblocks this PR

flash1293 commented:
@alex-gron Thanks for cycling back to this - your points match with my understanding. Let spin off the discussion of whether to put it into a separate table. As you mention this is separate and can be changed after the fact.

aaronsteers commented:
> Thanks for your comments @aaronsteers - I addressed everything and it seems like it's ready to merge but leaving it in place in case you want to take a final look.

Thanks! Looks great. For my part, at least, feel free to merge when ready.
